src/librustc_mir/borrow_check/mod.rs

```diff
@@ -708,6 +702,66 @@ impl InitializationRequiringAction {
 }
 
 impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
+
+    fn visit_terminator_drop(
+        &mut self,
+        loc: Location,
+        term: &Terminator<'tcx>,
+        flow_state: &Flows<'cx, 'gcx, 'tcx>,
+        drop_place: &Place<'tcx>,
+        span: Span,
+    ) {
+        let ty = drop_place.ty(self.mir, self.tcx).to_ty(self.tcx);
+        match ty.sty {
+            // When a struct is being dropped, we need to check whether it has a
+            // destructor, if it does, then we can call it, if it does not then we
+            // need to check the individual fields instead.
+            // See #47703.
+            ty::TyAdt(def, substs) if def.is_struct() && !def.has_dtor(self.tcx) => {
+                for (index, field) in def.all_fields().enumerate() {
+                    let field_ty = field.ty(self.tcx, substs);
```

Comment:
Sadly, before landing we're going to have to normalize associated types here. I have to dig a bit to see what's the most convenient way to do that. Probably we can erase regions, actually, which might simplify our lives a bit.